persistence/internal/providertest/offsetstore/transaction.go

```diff
@@ -0,0 +1,198 @@
+
+							return nil
+						},
+					)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+				})
+			})
+
+			ginkgo.When("an OCC conflict occurs", func() {
+				table.DescribeTable(
+					"it does not update the offset",
+					func(conflictingCurrentOffset, nextOffset offsetstore.Offset) {
+						// Update the current offset twice so that it's up to 2.
+						// Otherwise we can't test for 1 as a too-low value.
+						c := offsetstore.Offset(0)
+						n := offsetstore.Offset(1)
+
+						saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+						c = n
+						n += n
+						saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+						err := persistence.WithTransaction(
+							tc.Context,
+							dataStore,
+							func(tx persistence.ManagedTransaction) error {
+								err := tx.SaveOffset(
+									tc.Context,
+									"<source-app-key>",
+									conflictingCurrentOffset,
+									nextOffset,
+								)
+								gomega.Expect(err).To(gomega.Equal(offsetstore.ErrConflict))
+
+								return nil // let the transaction commit despite error
+							},
+						)
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+						actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+						gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+					},
+					table.Entry("zero and one", 0, 1),
+					table.Entry("too low pair", 1, 2),
+					table.Entry("too high pair", 99, 100),
+				)
+			})
+
+			ginkgo.When("the transaction is rolled-back", func() {
+				ginkgo.It("does not save the offset", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+
+					err := common.WithTransactionRollback(
+						tc.Context,
+						dataStore,
+						func(tx persistence.ManagedTransaction) error {
+							return tx.SaveOffset(
+								tc.Context,
+								"<source-app-key>",
+								c,
+								n,
+							)
+						},
+					)
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).Should(gomega.BeNumerically("==", c))
+				})
+			})
+
+			ginkgo.When("used in concurrent transactions", func() {
+				ginkgo.It("saves offsets concurrently", func() {
+					var g sync.WaitGroup
+
+					fn := func(ak string, c, n offsetstore.Offset) {
+						defer ginkgo.GinkgoRecover()
+						defer g.Done()
+
+						err := persistence.WithTransaction(
+							tc.Context,
+							dataStore,
+							func(tx persistence.ManagedTransaction) error {
+								return tx.SaveOffset(tc.Context, ak, c, n)
+							},
+						)
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					}
+
+					g.Add(3)
+					go fn("<source-app1-key>", 0, 1)
```

Comment:
Maybe use distinct "next" values for each app here, so we are sure they don't get swapped somehow.